File: app/src/main/java/com/termux/terminal/SessionManager.java
package com.termux.terminal;

import com.termux.NyxActivity;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import nyx.constants.Constant;

public final class SessionManager {
    /**
     * List of Sessions
     */
    public static final List<TerminalSession> sessions = new ArrayList<>(1);

    /**
     * Remove a TerminalSession.
     */
    private static int removeTerminalSession(final TerminalSession sessionToRemove) {
        sessionToRemove.finishIfRunning();
        SessionManager.sessions.remove(sessionToRemove);
        return SessionManager.sessions.size() - 1;
    }

    public static void addNewSession(final boolean isFailSafe) {
        SessionManager.sessions.add(SessionManager.createTerminalSession(isFailSafe));
        NyxActivity.console.attachSession(SessionManager.sessions.size() - 1);
    }

    private static TerminalSession createTerminalSession(final boolean isFailSafe) {
        final boolean failsafeCheck = isFailSafe || !new File(Constant.USR_DIR).exists();
        return new TerminalSession(failsafeCheck);
    }

    public static void removeFinishedSession(final TerminalSession finishedSession) { // Return pressed with finished session - remove it.
        final int index = SessionManager.removeTerminalSession(finishedSession);
        if (-1 == index) System.exit(0);
        else NyxActivity.console.attachSession(index);
    }

    public static void removeAll() {
        for (final TerminalSession i : SessionManager.sessions)
            SessionManager.removeFinishedSession(i);
    }
}

File: app/src/main/java/com/termux/terminal/SessionManager.kt
package com.termux.terminal

import com.termux.NyxActivity
import com.termux.data.ConfigManager.FILES_DIR_PATH
import com.termux.data.console
import java.io.File

object SessionManager {
    /**
     * List of Sessions
     */
    val sessions: MutableList<TerminalSession> = mutableListOf()


    /**
     * Remove a TerminalSession.
     */
    private fun removeTerminalSession(sessionToRemove: TerminalSession): Int {
        sessionToRemove.finishIfRunning()
        sessions.remove(sessionToRemove)
        return sessions.size - 1
    }

    fun addNewSession(isFailSafe: Boolean) {
        sessions.add(createTerminalSession(isFailSafe))
        console.attachSession(sessions.size - 1)
    }

    private fun createTerminalSession(isFailSafe: Boolean): TerminalSession {
        val failsafeCheck = isFailSafe || !File("$FILES_DIR_PATH/usr").exists()
        val newTerminalSession = TerminalSession(failsafeCheck)
        return newTerminalSession
    }

    fun removeFinishedSession(finishedSession: TerminalSession) {
        // Return pressed with finished session - remove it.
        val index = removeTerminalSession(finishedSession)
        if (index == -1) (console.context as NyxActivity).destroy()
        else console.attachSession(index)

    }
}
